libraries/AP_CINS/AP_CINS.cpp

```diff
@@ -0,0 +1,801 @@
+    heapVars.A23 = Gamma.R() * heapVars.GammaInv.A().a21();
+    heapVars.A31 = - Matrix3F::skew_symmetric(heapVars.GammaInv.W2()) * Gamma.R();
+    heapVars.A32 = Gamma.R() * heapVars.GammaInv.A().a12();
+    heapVars.A33 = Gamma.R() * heapVars.GammaInv.A().a22();
+
+    // Implement M(t+) = A M(t)
+    heapVars.M1Gyr = state.gyr_bias_gain_mat.rot;
+    heapVars.M2Gyr = state.gyr_bias_gain_mat.vel;
+    heapVars.M3Gyr = state.gyr_bias_gain_mat.pos;
+    state.gyr_bias_gain_mat.rot = heapVars.A11 * heapVars.M1Gyr;
+    state.gyr_bias_gain_mat.rot += heapVars.A12 * heapVars.M2Gyr;
+    state.gyr_bias_gain_mat.rot += heapVars.A13 * heapVars.M3Gyr;
+    state.gyr_bias_gain_mat.vel = heapVars.A21 * heapVars.M1Gyr;
+    state.gyr_bias_gain_mat.vel += heapVars.A22 * heapVars.M2Gyr;
+    state.gyr_bias_gain_mat.vel += heapVars.A23 * heapVars.M3Gyr;
+    state.gyr_bias_gain_mat.pos = heapVars.A31 * heapVars.M1Gyr;
+    state.gyr_bias_gain_mat.pos += heapVars.A32 * heapVars.M2Gyr;
+    state.gyr_bias_gain_mat.pos += heapVars.A33 * heapVars.M3Gyr;
+
+    heapVars.M1Acc = state.acc_bias_gain_mat.rot;
+    heapVars.M2Acc = state.acc_bias_gain_mat.vel;
+    heapVars.M3Acc = state.acc_bias_gain_mat.pos;
+    state.acc_bias_gain_mat.rot = heapVars.A11 * heapVars.M1Acc;
+    state.acc_bias_gain_mat.rot += heapVars.A12 * heapVars.M2Acc;
+    state.acc_bias_gain_mat.rot += heapVars.A13 * heapVars.M3Acc;
+    state.acc_bias_gain_mat.vel = heapVars.A21 * heapVars.M1Acc;
+    state.acc_bias_gain_mat.vel += heapVars.A22 * heapVars.M2Acc;
+    state.acc_bias_gain_mat.vel += heapVars.A23 * heapVars.M3Acc;
+    state.acc_bias_gain_mat.pos = heapVars.A31 * heapVars.M1Acc;
+    state.acc_bias_gain_mat.pos += heapVars.A32 * heapVars.M2Acc;
+    state.acc_bias_gain_mat.pos += heapVars.A33 * heapVars.M3Acc;
+}
+
+
+void AP_CINS::saturate_bias(Vector3F& bias_correction, const Vector3F& current_bias, const ftype& saturation, const ftype& dt) const
+{
+    // Ensure that no part of the bias exceeds the saturation limit
+    if (abs(bias_correction.x) > 1E-8) {
+        bias_correction.x *= MIN(1., (saturation - abs(current_bias.x)) / (dt * abs(bias_correction.x)));
+    }
+    if (abs(bias_correction.y) > 1E-8) {
+        bias_correction.y *= MIN(1., (saturation - abs(current_bias.y)) / (dt * abs(bias_correction.y)));
+    }
+    if (abs(bias_correction.z) > 1E-8) {
+        bias_correction.z *= MIN(1., (saturation - abs(current_bias.z)) / (dt * abs(bias_correction.z)));
+    }
+}
+
+
+Vector3F computeRotationCorrection(const Vector3F& v1, const Vector3F& v2, const ftype& gain, const ftype& dt)
+{
+    // The correction is given by Omega = - skew(v1) * v2 * gain, where v2 = R v0 for some const. v0.
+    // It is applied as dR/dt = skew(Omega) R.
+    // The key point is that Omega changes R which changes Omega.
+    // We use a first order expansion to capture this effect here.
+
+    const Vector3F omega1 = - Matrix3F::skew_symmetric(v1) * v2 * gain;
```

Comment:
```suggestion
    const auto skewed_v1 = Matrix3F::skew_symmetric(v1);
    const Vector3F omega1 = - skewed_v1 * v2 * gain;
```

and use this elsewhere.

similarly for skewed omega1